internal/core/database_api/athena/driver/athena.go

```
@@ -0,0 +1,207 @@
+		if err != nil {
+			output.Status = models.QueryFailed
+			output.ErrorMessage = "DoQuery failed" // simple error message for lambda caller
+		}
+	}()
+
+	query := awsathena.NewAthenaQuery(client, input.DatabaseName, input.SQL, s3Path)
+	err = query.Run()
+	if err != nil {
+		return output, err
+	}
+	err = query.Wait()
+	if err != nil {
+		return output, err
+	}
+
+	output.Status = models.QuerySucceeded
+
+	err = collectResults(query.QueryResult, (*models.GetQueryResultsOutput)(output), nil)
+	if err != nil {
+		return output, err
+	}
+
+	return output, nil
+}
+
+func StartQuery(client athenaiface.AthenaAPI, input *models.StartQueryInput, s3Path *string) (*models.StartQueryOutput, error) {
+	output := &models.StartQueryOutput{}
+
+	var err error
+	defer func() {
+		if err != nil {
+			output.Status = models.QueryFailed
+			output.ErrorMessage = "StartQuery failed" // simple error message for lambda caller
+		}
+	}()
+
+	query := awsathena.NewAthenaQuery(client, input.DatabaseName, input.SQL, s3Path)
+	err = query.Run()
+	if err != nil {
+		return output, err
+	}
+
+	output.QueryID = *query.StartExecutionOutput.QueryExecutionId
+
+	time.Sleep(minimalQueryWait) // give query opportunity to finish and avoid a later polling step
+
+	executionStatus, done, err := query.Done()
+	if err != nil {
+		return output, err
+	}
+	output.Status = getQueryStatus(executionStatus)
+	if done { // fill results
+		if output.Status == models.QuerySucceeded {
+			err = getQueryResults(query.Client, executionStatus, &output.GetQueryResultsOutput, nil, input.MaxResults)
+			if err != nil {
+				return output, err
+			}
+		}
+	}
+	return output, nil
+}
+
+func GetQueryStatus(client athenaiface.AthenaAPI, input *models.GetQueryStatusInput) (*models.GetQueryStatusOutput, error) {
+	output := &models.GetQueryStatusOutput{}
+
+	var err error
+	defer func() {
+		if err != nil {
+			output.ErrorMessage = "GetQueryStatus failed" // simple error message for lambda caller
+		}
+	}()
```

I don't understand the purpose of having an "Error" string in the response but also have the lambda return an error. 
* If we want to return an error message, you can just use the "genericapi` errors that allow us to include detailed messages
* If we want to return an error message in the response but don't cause the lambda to fail, this method shouldn't return an error at all in this code.

---

Let me review all the error scenarios and see if I can use the genericapi error for them.

---

I've updated the error handling.